### PR TITLE
core.internal.elf: Improve usability

### PR DESCRIPTION
**Pull request by kinke**
Paving the way for a revamped `rt.sections_android` coming up next (ldc-developers/druntime#178).

### CODE REVIEW

File: src/core/internal/elf/io.d [hunk 1]
```diff
@@ -285,3 +315,29 @@ private struct MMapRegion(T)
...
+        {
+            auto beg = exe.baseAddress + offset;
+            printf("%p - %p\n", beg, beg + sectionHeader.shdr.sh_size);
+        }
+        else
+        {
+            printf("not mapped into memory\n");
+        }
+    }
+}
```
Comment (kinke) [inline src/core/internal/elf/io.d]: E.g.:
```
DSO name: runtime/druntime-test-runner-shared
    path: /home/martin/build-ldc/runtime/druntime-test-runner-shared
    base: 0x5589831e6000
DSO name: linux-vdso.so.1
    path: [vdso]
    base: 0x7ffdeb9f7000
DSO name: /home/martin/build-ldc/lib/libdruntime-ldc-unittest-shared.so.91
    path: /home/martin/build-ldc/lib/libdruntime-ldc-unittest-shared.so.2.0.91
    base: 0x7ff21166b000
DSO name: /lib/x86_64-linux-gnu/libgcc_s.so.1
    path: /lib/x86_64-linux-gnu/libgcc_s.so.1
    base: 0x7ff211453000
DSO name: /lib/x86_64-linux-gnu/libc.so.6
    path: /lib/x86_64-linux-gnu/libc-2.27.so
    base: 0x7ff211062000
DSO name: /lib64/ld-linux-x86-64.so.2
    path: /lib/x86_64-linux-gnu/ld-2.27.so
    base: 0x7ff211b53000
DSO name: /lib/x86_64-linux-gnu/libm.so.6
    path: /lib/x86_64-linux-gnu/libm-2.27.so
    base: 0x7ff210cc4000
DSO name: /lib/x86_64-linux-gnu/libpthread.so.0
    path: /lib/x86_64-linux-gnu/libpthread-2.27.so
    base: 0x7ff210aa5000
DSO name: /lib/x86_64-linux-gnu/librt.so.1
    path: /lib/x86_64-linux-gnu/librt-2.27.so
    base: 0x7ff21089d000
DSO name: /lib/x86_64-linux-gnu/libdl.so.2
    path: /lib/x86_64-linux-gnu/libdl-2.27.so
    base: 0x7ff210699000
```

File: src/core/internal/elf/dl.d [hunk 2]
```diff
@@ -160,3 +213,20 @@ else
...
+    {
+        const name = object.name();
+        assert(name.length);
+        const path = object.getPath(buffer);
+
+        printf("DSO name: %s\n", name.ptr);
+        printf("    path: %s\n", path ? path.ptr : "");
+        printf("    base: %p\n", object.baseAddress);
+    }
+}
```
Comment (kinke) [inline src/core/internal/elf/dl.d]: E.g.:
```
DSO name: runtime/druntime-test-runner-shared
    path: /home/martin/build-ldc/runtime/druntime-test-runner-shared
    base: 0x5589831e6000
DSO name: linux-vdso.so.1
    path: [vdso]
    base: 0x7ffdeb9f7000
DSO name: /home/martin/build-ldc/lib/libdruntime-ldc-unittest-shared.so.91
    path: /home/martin/build-ldc/lib/libdruntime-ldc-unittest-shared.so.2.0.91
    base: 0x7ff21166b000
DSO name: /lib/x86_64-linux-gnu/libgcc_s.so.1
    path: /lib/x86_64-linux-gnu/libgcc_s.so.1
    base: 0x7ff211453000
DSO name: /lib/x86_64-linux-gnu/libc.so.6
    path: /lib/x86_64-linux-gnu/libc-2.27.so
    base: 0x7ff211062000
DSO name: /lib64/ld-linux-x86-64.so.2
    path: /lib/x86_64-linux-gnu/ld-2.27.so
    base: 0x7ff211b53000
DSO name: /lib/x86_64-linux-gnu/libm.so.6
    path: /lib/x86_64-linux-gnu/libm-2.27.so
    base: 0x7ff210cc4000
DSO name: /lib/x86_64-linux-gnu/libpthread.so.0
    path: /lib/x86_64-linux-gnu/libpthread-2.27.so
    base: 0x7ff210aa5000
DSO name: /lib/x86_64-linux-gnu/librt.so.1
    path: /lib/x86_64-linux-gnu/librt-2.27.so
    base: 0x7ff21089d000
DSO name: /lib/x86_64-linux-gnu/libdl.so.2
    path: /lib/x86_64-linux-gnu/libdl-2.27.so
    base: 0x7ff210699000
```

File: src/core/internal/elf/io.d [hunk 3]
```diff
@@ -134,41 +135,70 @@ template ElfIO(Elf_Ehdr, Elf_Shdr, ubyte ELFCLASS)
...
+            }
+            return false;
         }
+    }
 
-        /**
-         * Tries to find the index of the section with the specified name.
-         * Returns: -1 if not found, otherwise 0-based section index.
-         */
-        size_t findSectionIndexByName(const(char)[] sectionName) const
```
Comment (kinke) [inline src/core/internal/elf/io.d]: This function has been removed. These modules have just been added with v2.091 (#2330); at the time I originally worked on this (~1.5 years ago), there was no `move` in druntime (only in Phobos), so I've resorted to this little helper for `findSectionHeaderByName()`.

File: src/core/internal/elf/dl.d [hunk 4]
```diff
@@ -160,3 +213,20 @@ else
...
+unittest
+{
+    import core.stdc.stdio;
+
+    char[512] buffer = void;
+    foreach (object; SharedObjects)
+    {
+        const name = object.name();
+        assert(name.length);
+        const path = object.getPath(buffer);
```
Comment (kinke) [inline src/core/internal/elf/dl.d]: I've originally asserted that `path` isn't null (no error), but that failed for some Linux CI systems for LDC. I haven't bothered looking into what the issue is.

File: src/core/internal/elf/dl.d [hunk 5]
```diff
@@ -112,6 +120,42 @@ struct SharedObject
...
+     * Returns: The filled slice (null-terminated), or null if an error occurs.
+     */
+    char[] getPath(size_t N)(ref char[N] buffer) const
+    if (N > 1)
+    {
+        import core.stdc.stdio, core.stdc.string, core.sys.posix.unistd;
+
+        char[N + 128] lineBuffer = void;
+
+        snprintf(lineBuffer.ptr, lineBuffer.length, "/proc/%d/maps", getpid());
```
Comment (kubo39) [inline src/core/internal/elf/dl.d]: In FreeBSD, using `kinfo_getvmmap` seems more suitable, this works even if /proc is not mounted.
https://www.freebsd.org/cgi/man.cgi?query=kinfo_getvmmap

File: src/core/internal/elf/dl.d [hunk 6]
```diff
@@ -148,7 +148,16 @@ struct SharedObject
 
 private @nogc nothrow:
 
-version (linux)
+version (CRuntime_Bionic)
```
[the remark this answers is not in the document]
Comment (kinke) [reply src/core/internal/elf/dl.d]: Heh, I was a bit lazy and took this workaround from the guy who tested this stuff. I'll try to get ahold of the Android NDK headers and check,

File: src/core/internal/elf/dl.d (hunk 5 above)
Comment (kinke) [reply src/core/internal/elf/dl.d]: I'm not sure that's worth it (incl. the `libutil` dependency); I've just tried to write this in a portable way (e.g., no /proc/**self**/maps etc.).

**Comment by kinke** [comment]
> Since the unittests output something, can they be made integration tests ?

You mean a test with Makefile etc. in the `test` dir? My primary motivation was to add examples of how to use these modules (and test that the code is not totally broken), without digging into platform specifics for thorough segment/section checks. That's why I'd prefer to keep the tests directly in these modules, so that people find them.

File: src/core/internal/elf/dl.d (hunk 6 above)
[the remark this answers is not in the document]
Comment (kinke) [reply src/core/internal/elf/dl.d]: By the looks of it, `__USE_GNU` etc. is defined for most Linux C libs in https://github.com/dlang/druntime/blob/master/src/core/sys/posix/config.d (notable exception: `version(CRuntime_Musl)`), incl. being `false` for Android/Bionic. The uncoditional enums in core.sys.**linux**.config (publicly importing core.sys.**posix**.config but 'overriding' with unconditional enums, apparently from glibc) seem wrong and obsolete. I guess the proper fix would be to make sure all expected enums are already set up in the Posix module, for all Linux C libs.

[the remark this answers is not in the document]
Comment (kinke) [reply src/core/internal/elf/dl.d]: That looks like a dangerous change though (I've just seen that `__USE_GNU` is false for glibc in ...posix.config!), so I don't want to add it in this PR. I'll add a TODO comment here.

**Pull request merged**